src/main.rs

```diff
@@ -189,47 +187,68 @@ fn main() {
-                        .map_err(|err| io::Error::new(io::ErrorKind::Other, err))
-                })
-                .map_err(move |err| {
-                    task_warn!("TLS handshake failed"; "error" => format!("{}", err));
-                })
-                .and_then(move |tls_socket| {
-                    // the Bridge::create returns a future that resolves to a Bridge object. The
-                    // Bridge object is *also* a future that we want to wait on, so we use `flatten()`
-                    bridge::Bridge::create(new_handle, cloned_url, tls_socket, irc_server_name, ctx)
-                        .map(|bridge| bridge.into_future())
-                        .flatten()
-                        .map_err(
-                            |err| task_warn!("Unhandled IO error"; "error" => format!("{}", err)),
-                        )
-                })
-                .then(|r| {
-                    task_info!("Finished");
-                    r
-                });
+            let spawn_fut = future::lazy(move |_cx: &mut Context| async move {
+                debug!(ctx.logger.as_ref(), "Accepted connection");
+
+                let tls_socket = acceptor
+                    .accept(tcp_stream)
+                    .await
+                    .map_err(|err| {
+                        task_warn!(ctx, "TLS handshake failed"; "error" => format!("{}", err));
+                        io::Error::new(io::ErrorKind::Other, err);
+                    })
+                    .unwrap();
+
+                let mut bridge =
+                    bridge::Bridge::create(cloned_url, tls_socket, irc_server_name, ctx.clone())
+                        .await
+                        .unwrap();
+
+                // TODO: I belive this is what taskedfutures did (except it was previously spawned
+                // off inside Bridge::create). I will come back and make sure this is correct
+                // later.
+                loop {
+                    if let Err(e) = bridge.poll_irc().await {
+                        task_warn!(ctx, "Encounted error while polling IRC connection"; "error" => format!{"{}", e});
+                        break;
+                    }
+                    if let Err(e) = bridge.poll_matrix().await {
+                        task_warn!(ctx, "Encounted error while polling matrix connection"; "error" => format!{"{}", e});
+                        break;
+                    }
+                }
+
+                task_info!(ctx, "Finished");
+            });
 
             // We spawn the future off, otherwise we'd block the stream of incoming connections.
             // This is what causes the future to be in its own chain.
-            handle.spawn(future);
+            //
+            // These are spawn_local due to it not requiring spawn_fut to be Send, but it could
+            // possibly be `spawn` in the future after changing trait bounds.
+            tokio::task::spawn_local(spawn_fut);
```

Comment:
[This function panics if called outside of a local task set.](https://docs.rs/tokio/0.2.22/tokio/task/fn.spawn_local.html) You can't use it like this.

Reply:
I'm guessing this also got lost in my not-actually-submitted previous review.

Reply:
This means that if `spawn_local` is called outside of a task it will panic? So `tokio::spawn` would be required, and therefore everything must be `Send`?

Reply:
No. Everything is inside a task with `#[tokio::main]`. What this is talking about it `LocalTaskSet`. Please look at the linked documentation (and the one for `LocalTaskSet`) if you want to know more. We probably should be making everything `Send` and not use `LocalTaskSet` though.